src/test/ui/issues/issue-50480.stderr

```diff
@@ -4,13 +4,14 @@ error[E0412]: cannot find type `NotDefined` in this scope
 LL | struct Foo(NotDefined, <i32 as Iterator>::Item, Vec<i32>, String);
    |            ^^^^^^^^^^ not found in this scope
 
-error[E0277]: the trait bound `i32: std::iter::Iterator` is not satisfied
+error[E0277]: `i32` is not an iterator
   --> $DIR/issue-50480.rs:13:24
    |
 LL | struct Foo(NotDefined, <i32 as Iterator>::Item, Vec<i32>, String);
-   |                        ^^^^^^^^^^^^^^^^^^^^^^^ `i32` is not an iterator; maybe try calling `.iter()` or a similar method
+   |                        ^^^^^^^^^^^^^^^^^^^^^^^ `i32` is not an iterator
    |
    = help: the trait `std::iter::Iterator` is not implemented for `i32`
+   = note: if you want to iterate between `0` until a value `end`, use the range syntax: `0..end`
```

Comment:
That was my thinking.

Reply:
This would also be fixed by marking all the spans part of a for loop as being part of a desugaring context, which we can filter on in `rustc_on_unimplemented` and would improve the situation for many of this diagnostics that only apply when using a binding and not in type signatures.

Reply:
Ah, that would be good!